### PR TITLE
mgr/cephadm: store contianer registry credentials in config-key

### PR DESCRIPTION
**Pull request by Daniel-Pivonka**
store credentials in a more secure place 

Fixes: https://tracker.ceph.com/issues/53269
Signed-off-by: Daniel Pivonka <dpivonka@redhat.com>

### CODE REVIEW

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -4859,9 +4859,9 @@ def command_bootstrap(ctx):
         prepare_ssh(ctx, cli, wait_for_mgr_restart)
 
     if ctx.registry_url and ctx.registry_username and ctx.registry_password:
-        cli(['config', 'set', 'mgr', 'mgr/cephadm/registry_url', ctx.registry_url, '--force'])
-        cli(['config', 'set', 'mgr', 'mgr/cephadm/registry_username', ctx.registry_username, '--force'])
-        cli(['config', 'set', 'mgr', 'mgr/cephadm/registry_password', ctx.registry_password, '--force'])
+        cli(['config-key', 'set', 'mgr/cephadm/registry_url', ctx.registry_url])
+        cli(['config-key', 'set', 'mgr/cephadm/registry_username', ctx.registry_username])
+        cli(['config-key', 'set', 'mgr/cephadm/registry_password', ctx.registry_password])
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:4864]: can we just store a json string?

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -277,6 +277,7 @@ class CephadmOrchestrator(orchestrator.Orchestrator, MgrModule,
             default='*',
             desc='PlacementSpec describing on which hosts to manage /etc/ceph/ceph.conf',
         ),
+        # not used anymore
```
Comment (liewegas) [inline src/pybind/mgr/cephadm/module.py:280]: I think we should remove these options entirely.  If I'm reading the code properly, the migration ``config rm`` calls should still work.

Comment (Daniel-Pivonka) [reply src/pybind/mgr/cephadm/module.py:280]: i wanted to do that but i was having trouble getting the old values during the migration.

once the new mgr takes over during an upgrade and with those removed
`registry_url = self.mgr.get_module_option('registry_url')`
causes
`RuntimeError: Config option 'registry_url' is not in CephadmOrchestrator.MODULE_OPTIONS`


and 
```
ret, registry_url, err = self.mgr.mon_command({
                'prefix': 'config get',
                'who': 'mgr',
                'key': 'mgr/cephadm/registry_url',
            })
```
returns a return code of `-2` and no value for registry_url


additionally trying to run 

`ceph config get mgr mgr/cephadm/registry_url`

returns `Error ENOENT:
`

File: src/pybind/mgr/cephadm/migrations.py [hunk 3]
```diff
@@ -259,6 +263,39 @@ def migrate_nfs_spec(self, service_id: str, pool: str, ns: Optional[str]) -> Non
...
+
+            registry_credentials = {'url': registry_url, 'username': registry_username, 'password': registry_password}
+            self.mgr.set_store('registry_credentials', json.dumps(registry_credentials))
+
+            self.mgr.set_module_option('registry_url', None)
+            self.mgr.check_mon_command({
+                'prefix': 'config rm',
+                'who': 'mgr',
+                'key': 'mgr/cephadm/registry_url',
+            })
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/migrations.py:280]: think this is redundant. See 

https://github.com/ceph/ceph/blob/8713c241a1bb1a6a41dea43e922e5124193fcf01/src/mgr/PyModule.cc#L199-L204

File: src/pybind/mgr/cephadm/serve.py [hunk 4]
```diff
@@ -1292,19 +1289,15 @@ def _get_container_image_info(self, image_name: str) -> ContainerInspectInfo:
         return r
 
     # function responsible for logging single host into custom registry
-    def _registry_login(self, host: str, url: Optional[str], username: Optional[str], password: Optional[str]) -> Optional[str]:
-        self.log.debug(f"Attempting to log host {host} into custom registry @ {url}")
+    def _registry_login(self, host: str, registry_json: str) -> Optional[str]:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/serve.py:1292]:
```suggestion
    def _registry_login(self, host: str, registry_json: Dict[str,str]) -> Optional[str]:
```
Let's avoid serializing and deserializing things internally.

File: src/pybind/mgr/cephadm/serve.py [hunk 5]
```diff
@@ -1292,19 +1289,15 @@ def _get_container_image_info(self, image_name: str) -> ContainerInspectInfo:
...
         # want to pass info over stdin rather than through normal list of args
-        args_str = json.dumps({
-            'url': url,
-            'username': username,
-            'password': password,
-        })
         out, err, code = self._run_cephadm(
             host, 'mon', 'registry-login',
-            ['--registry-json', '-'], stdin=args_str, error_ok=True)
+            ['--registry-json', '-'], stdin=registry_json, error_ok=True)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/serve.py:1298]:
```suggestion
            ['--registry-json', '-'], stdin=json.dumps(registry_json), error_ok=True)
```

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

**Comment by sebastian-philipp** [comment]
While you're here, https://docs.ceph.com/en/latest/cephadm/install/#further-information-about-cephadm-bootstrap is bad as well. Can we deprecate `cephadm bootstrap --registry-password`?

File: src/pybind/mgr/cephadm/module.py (hunk 2 above)
Comment (liewegas) [reply src/pybind/mgr/cephadm/module.py:280]: I think I should add a ``--force`` to ``config get``... WDYT @sebastian-philipp ?

Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:280]: Not yet made my mind about this one. IMO this is an extraordinary edge case. Is it worth building extra code for this?

File: src/pybind/mgr/cephadm/migrations.py (hunk 3 above)
Comment (Daniel-Pivonka) [reply src/pybind/mgr/cephadm/migrations.py:280]: without this the values do not get removed from config dump

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
```
2021-11-29T16:40:23.637 DEBUG:teuthology.orchestra.run.smithi026:> sudo /home/ubuntu/cephtest/cephadm --image docker.io/ceph/ceph:v16.2.4 shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid 65fb89d2-5130-11ec-8c2d-001a4aab830c -- bash -c 'ceph config get mgr mgr/cephadm/migration_current | grep 3'
2021-11-29T16:40:24.741 INFO:journalctl@ceph.mon.smithi026.smithi026.stdout:Nov 29 16:40:24 smithi026 ceph-mon[106998]: pgmap v356: 129 pgs: 129 active+clean; 214 MiB data, 2.7 GiB used, 712 GiB / 715 GiB avail; 255 B/s rd, 0 op/s
2021-11-29T16:40:25.037 INFO:journalctl@ceph.mon.smithi119.smithi119.stdout:Nov 29 16:40:24 smithi119 ceph-mon[74838]: pgmap v356: 129 pgs: 129 active+clean; 214 MiB data, 2.7 GiB used, 712 GiB / 715 GiB avail; 255 B/s rd, 0 op/s
2021-11-29T16:40:25.968 DEBUG:teuthology.orchestra.run:got remote process result: 1
2021-11-29T16:40:25.969 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_a0bc318848afe71da1fbf34e1e1b44738d8e3d6d/teuthology/run_tasks.py", line 91, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_a0bc318848afe71da1fbf34e1e1b44738d8e3d6d/teuthology/run_tasks.py", line 70, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_caa4f4125affaece3c1649c79f1b56fa72f9867e/qa/tasks/cephadm.py", line 1101, in shell
    extra_cephadm_args=args)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_caa4f4125affaece3c1649c79f1b56fa72f9867e/qa/tasks/cephadm.py", line 47, in _shell
    **kwargs
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_a0bc318848afe71da1fbf34e1e1b44738d8e3d6d/teuthology/orchestra/remote.py", line 509, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_a0bc318848afe71da1fbf34e1e1b44738d8e3d6d/teuthology/orchestra/run.py", line 455, in run
    r.wait()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_a0bc318848afe71da1fbf34e1e1b44738d8e3d6d/teuthology/orchestra/run.py", line 161, in wait
    self._raise_for_status()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_a0bc318848afe71da1fbf34e1e1b44738d8e3d6d/teuthology/orchestra/run.py", line 183, in _raise_for_status
    node=self.hostname, label=self.label
teuthology.exceptions.CommandFailedError: Command failed on smithi026 with status 1: "sudo /home/ubuntu/cephtest/cephadm --image docker.io/ceph/ceph:v16.2.4 shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid 65fb89d2-5130-11ec-8c2d-001a4aab830c -- bash -c 'ceph config get mgr mgr/cephadm/migration_current | grep 3'"
```

https://pulpito.ceph.com/swagner-2021-11-29_16:02:52-orch:cephadm-wip-swagner3-testing-2021-11-29-1155-distro-default-smithi/6533675 

See 

https://github.com/ceph/ceph/blob/174ad996c15ee7d50b3c74107946ffc0204b2139/qa/suites/orch/cephadm/mgr-nfs-upgrade/4-final.yaml#L10

**Pull request merged**